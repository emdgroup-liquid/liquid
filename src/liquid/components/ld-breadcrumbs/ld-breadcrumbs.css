:host,
.ld-breadcrumbs {
  --ld-crumb-icon-gap: 0.5em;
  --ld-crumb-gap: 0.6em;

  line-height: 1.5;
}

.ld-breadcrumbs > ol {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  padding: 0;
  margin: 0;
}
