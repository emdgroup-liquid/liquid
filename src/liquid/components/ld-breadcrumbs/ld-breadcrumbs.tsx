import { Component, h, Element } from '@stencil/core'

/**
 * @part list - Breadcrumbs list
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */

@Component({
  tag: 'ld-breadcrumbs',
  styleUrl: 'ld-breadcrumbs.css',
  shadow: true,
})
export class LdBreadcrumbs {
  @Element() el: HTMLElement

  private observer: MutationObserver

  private updateCurrent = () => {
    const crumbs = this.el.querySelectorAll('ld-crumb')
    if (!crumbs.length) return

    crumbs.forEach((crumb) => {
      crumb.current = undefined
    })
    crumbs[crumbs.length - 1].current = true
  }

  componentDidLoad() {
    this.observer = new MutationObserver(this.updateCurrent)
    this.observer.observe(this.el, {
      subtree: true,
      childList: true,
      attributes: false,
    })

    this.updateCurrent()
  }

  disconnectedCallback() {
    /* istanbul ignore if */
    if (this.observer) this.observer.disconnect()
  }

  render() {
    return (
      <nav aria-label="Breadcrumbs" class="ld-breadcrumbs">
        <ol class="ld-breadcrumbs__list" part="list">
          <slot></slot>
        </ol>
      </nav>
    )
  }
}
