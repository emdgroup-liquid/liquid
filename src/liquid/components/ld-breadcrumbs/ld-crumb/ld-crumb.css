.ld-breadcrumbs .ld-link,
.ld-crumb__link::part(anchor) {
  --ld-link-gap: var(--ld-crumb-icon-gap);
  --ld-link-chevron-gap: var(--ld-crumb-gap);
  display: inline-flex;
}

.ld-breadcrumbs li:not(:last-of-type) .ld-link,
.ld-crumb__link:not(.ld-crumb__link--current)::part(anchor) {
  --ld-link-col: var(--ld-col-neutral-600);

  font-weight: normal;
  margin-right: calc(var(--ld-crumb-gap) + 0.5em);

  &:hover {
    --ld-link-col: var(--ld-thm-primary-hover);
  }

  &:focus:focus-visible {
    --ld-link-col: var(--ld-thm-primary-focus);
  }

  &:active {
    --ld-link-col: var(--ld-thm-primary-active);
  }
}

.ld-breadcrumbs li:last-of-type .ld-link,
.ld-crumb__link--current {
  cursor: default;
  pointer-events: none;
}
