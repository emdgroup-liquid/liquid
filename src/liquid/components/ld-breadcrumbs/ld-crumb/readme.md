---
eleventyNavigation:
  key: Crumb
  parent: Breadcrumbs
layout: layout.njk
title: Crumb
permalink: components/ld-breadcrumbs/ld-crumb/
---

# ld-crumb

The `ld-crumb` component is a subcomponent for `ld-breadcrumbs`.

Please refer to the [`ld-breadcrumbs` documentation](components/ld-breadcrumbs/) for usage examples.

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                              | Type               | Default     |
| -------- | --------- | -------------------------------------------------------- | ------------------ | ----------- |
| `href`   | `href`    | The URL that the hyperlink points to.                    | `string`           | `undefined` |
| `key`    | `key`     | for tracking the node's identity when working with lists | `string \| number` | `undefined` |
| `ref`    | `ref`     | reference to component                                   | `any`              | `undefined` |


## Shadow Parts

| Part     | Description     |
| -------- | --------------- |
| `"link"` | Breadcrumb link |


## Dependencies

### Depends on

- [ld-link](../../ld-link)

### Graph
```mermaid
graph TD;
  ld-crumb --> ld-link
  style ld-crumb fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

 
