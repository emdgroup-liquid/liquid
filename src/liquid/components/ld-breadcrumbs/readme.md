---
eleventyNavigation:
  key: Breadcrumbs
  parent: Components
layout: layout.njk
title: Breadcrumbs
permalink: components/ld-breadcrumbs/
---

<link rel="stylesheet" href="css_components/ld-breadcrumbs.css">
<link rel="stylesheet" href="css_components/ld-crumb.css">
<link rel="stylesheet" href="css_components/ld-link.css">
<link rel="stylesheet" href="css_components/ld-icon.css">

# ld-breadcrumbs

A breadcrumb trail consists of a list of links to the parent pages of the current page in hierarchical order. It helps users find their place within a website or web application. Breadcrumbs are often placed horizontally before a page's main content.

This component is meant to be used in conjunction with the [`ld-crumb`](./ld-crumb/) component.

## Examples

### Basic

{% example %}
<ld-breadcrumbs>
  <ld-crumb href="components/">Components</ld-crumb>
  <ld-crumb href="components/ld-breadcrumbs/">Breadcrumbs</ld-crumb>
</ld-breadcrumbs>

<!-- CSS component -->

<nav class="ld-breadcrumbs" aria-label="Breadcrumbs">
  <ol>
    <li>
      <a class="ld-link ld-link--chevron-end" href="components/">Components</a>
    </li>
    <li>
      <a class="ld-link" href="components/ld-breadcrumbs/" aria-current="page">Breadcrumbs</a>
    </li>
  </ol>
</nav>
{% endexample %}

### With icons

{% example %}
<ld-breadcrumbs>
  <ld-crumb href="components/">
    <ld-icon name="placeholder" size="sm"></ld-icon>
    Components
  </ld-crumb>
  <ld-crumb href="components/ld-breadcrumbs/">
    <ld-icon name="placeholder" size="sm"></ld-icon>
    Breadcrumbs
  </ld-crumb>
</ld-breadcrumbs>

<!-- CSS component -->

<nav class="ld-breadcrumbs" aria-label="Breadcrumbs">
  <ol>
    <li>
      <a class="ld-link ld-link--chevron-end" href="components/">
        <span class="ld-icon ld-icon--sm" role="presentation">
          <svg viewBox="0 0 24 24" fill="none">
            <rect x="1.5" y="1.5" width="21" height="21" rx="4.5" stroke="currentColor" stroke-width="3"/>
            <circle cx="12" cy="12" r="4.5" stroke="currentColor" stroke-width="3"/>
          </svg>
        </span>
        Components
      </a>
    </li>
    <li>
      <a class="ld-link" href="components/ld-breadcrumbs/" aria-current="page">
        <span class="ld-icon ld-icon--sm" role="presentation">
          <svg viewBox="0 0 24 24" fill="none">
            <rect x="1.5" y="1.5" width="21" height="21" rx="4.5" stroke="currentColor" stroke-width="3"/>
            <circle cx="12" cy="12" r="4.5" stroke="currentColor" stroke-width="3"/>
          </svg>
        </span>
        Breadcrumbs
      </a>
    </li>
  </ol>
</nav>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                              | Type               | Default     |
| -------- | --------- | -------------------------------------------------------- | ------------------ | ----------- |
| `key`    | `key`     | for tracking the node's identity when working with lists | `string \| number` | `undefined` |
| `ref`    | `ref`     | reference to component                                   | `any`              | `undefined` |


## Shadow Parts

| Part     | Description      |
| -------- | ---------------- |
| `"list"` | Breadcrumbs list |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
