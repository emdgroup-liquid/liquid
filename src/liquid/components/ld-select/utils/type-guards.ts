export const isLdOption = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionElement | HTMLLdOptionInternalElement =>
  ["LD-OPTION", "LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroup = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupElement | HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP", "LD-OPTGROUP-INTERNAL"].includes(
    (el as HTMLElement)?.tagName,
  );

export const isLdOptionInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionInternalElement =>
  ["LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroupInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP-INTERNAL"].includes((el as HTMLElement)?.tagName);

type HTMLLdOptInternal =
  | HTMLLdOptionInternalElement
  | HTMLLdOptgroupInternalElement;
export const isLdOptInternalHidden = (
  opt: HTMLLdOptInternal,
): opt is
  | (HTMLLdOptInternal & {
      hidden: true;
    })
  | (HTMLLdOptInternal & {
      filtered: true;
    }) => {
  return opt.hidden || opt.filtered;
};
