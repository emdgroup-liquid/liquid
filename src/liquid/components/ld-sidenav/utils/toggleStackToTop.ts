import { LdSidenavNavitem } from '../ld-sidenav-navitem/ld-sidenav-navitem'

export const toggleStackToTop = (el: HTMLElement, stacked: boolean) => {
  const outerSlider = el.closest('ld-sidenav-slider')
  if (!outerSlider) return

  const sidenavPaddingY =
    parseInt(
      window
        .getComputedStyle(outerSlider)
        .getPropertyValue('--ld-sidenav-padding-y')
    ) * 16

  // If not stacked, put everything back in place.
  if (!stacked) {
    el.style.removeProperty('--ld-slider-navitem-move-up')
    return
  }

  // Else, calculate how much vertical space is taken by stacked items above
  // and move the nav item up accordingly.
  let totalSpaceAbove = 0
  let totalSpaceOccupiedAbove = 0
  for (const elem of Array.from(el.parentElement.children)) {
    if (el === elem) break

    if (
      elem.tagName === 'LD-SIDENAV-SEPARATOR' ||
      (elem.tagName === 'LD-SIDENAV-NAVITEM' &&
        !['secondary', 'tertiary'].includes(
          (elem as unknown as LdSidenavNavitem).mode
        ))
    ) {
      totalSpaceOccupiedAbove += elem.getBoundingClientRect().height
    } else {
      totalSpaceAbove += sidenavPaddingY
    }
    totalSpaceAbove += elem.getBoundingClientRect().height
  }

  const spaceToMoveUp = totalSpaceAbove - totalSpaceOccupiedAbove
  el.style.setProperty('--ld-slider-navitem-move-up', `-${spaceToMoveUp}px`)
}
