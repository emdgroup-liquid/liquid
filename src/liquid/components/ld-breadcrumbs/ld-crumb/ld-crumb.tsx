import '../../../components' // type definitions for type checks and intelliSense
import { Component, h, Prop } from '@stencil/core'
import { getClassNames } from '../../../utils/getClassNames'

/**
 * @part link - Breadcrumb link
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-crumb',
  styleUrl: 'ld-crumb.css',
  shadow: true,
})
export class LdCrumb {
  /** @internal */
  @Prop() current?: boolean

  /** The URL that the hyperlink points to. */
  @Prop() href?: string

  render() {
    return (
      <li class="ld-crumb">
        <ld-link
          href={this.href}
          class={getClassNames([
            'ld-crumb__link',
            this.current && 'ld-crumb__link--current',
          ])}
          part="link"
          aria-current={this.current ? 'page' : undefined}
          chevron={this.current ? undefined : 'end'}
        >
          <slot></slot>
        </ld-link>
      </li>
    )
  }
}
