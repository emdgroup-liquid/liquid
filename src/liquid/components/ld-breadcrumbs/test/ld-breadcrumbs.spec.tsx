import { newSpecPage } from '@stencil/core/testing'
import { LdBreadcrumbs } from '../ld-breadcrumbs'
import { LdCrumb } from '../ld-crumb/ld-crumb'
import { getTriggerableMutationObserver } from '../../../utils/mutationObserver'

describe('ld-breadcrumbs', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdBreadcrumbs, LdCrumb],
      html: `<ld-breadcrumbs>
              <ld-crumb>foo</ld-crumb>
              <ld-crumb>bar</ld-crumb>
              <ld-crumb>qux</ld-crumb>
            </ld-breadcrumbs>`,
    })
    await page.waitForChanges()
    jest.advanceTimersByTime(0)

    expect(page.root).toMatchSnapshot()
  })

  it('updates current after addition of a crumb', async () => {
    const page = await newSpecPage({
      components: [LdBreadcrumbs, LdCrumb],
      html: `<ld-breadcrumbs>
              <ld-crumb>foo</ld-crumb>
              <ld-crumb>bar</ld-crumb>
              <ld-crumb>qux</ld-crumb>
            </ld-breadcrumbs>`,
    })
    await page.waitForChanges()
    jest.advanceTimersByTime(0)

    const crumb = page.doc.createElement('ld-crumb')
    crumb.innerHTML = 'baz'

    page.root.appendChild(crumb)
    getTriggerableMutationObserver().trigger([])
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })

  it('updates current after removal of current crumb', async () => {
    const page = await newSpecPage({
      components: [LdBreadcrumbs, LdCrumb],
      html: `<ld-breadcrumbs>
              <ld-crumb>foo</ld-crumb>
              <ld-crumb>bar</ld-crumb>
              <ld-crumb>qux</ld-crumb>
            </ld-breadcrumbs>`,
    })
    await page.waitForChanges()

    page.root.removeChild(page.root.querySelector('ld-crumb:last-of-type'))
    getTriggerableMutationObserver().trigger([])
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })

  it('Does not throw with no crumbs', async () => {
    const page = await newSpecPage({
      components: [LdBreadcrumbs, LdCrumb],
      html: `<ld-breadcrumbs></ld-breadcrumbs>`,
    })
    await page.waitForChanges()
    jest.advanceTimersByTime(0)

    expect(page.root).toMatchSnapshot()
  })
})
