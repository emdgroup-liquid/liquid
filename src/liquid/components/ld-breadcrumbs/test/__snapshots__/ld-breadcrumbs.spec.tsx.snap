// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-breadcrumbs renders 1`] = `
<ld-breadcrumbs>
  <mock:shadow-root>
    <nav aria-label="Breadcrumbs" class="ld-breadcrumbs">
      <ol class="ld-breadcrumbs__list" part="list">
        <slot></slot>
      </ol>
    </nav>
  </mock:shadow-root>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    foo
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    bar
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link aria-current="page" class="ld-crumb__link ld-crumb__link--current" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    qux
  </ld-crumb>
</ld-breadcrumbs>
`;

exports[`ld-breadcrumbs updates current after addition of a crumb 1`] = `
<ld-breadcrumbs>
  <mock:shadow-root>
    <nav aria-label="Breadcrumbs" class="ld-breadcrumbs">
      <ol class="ld-breadcrumbs__list" part="list">
        <slot></slot>
      </ol>
    </nav>
  </mock:shadow-root>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    foo
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    bar
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    qux
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link aria-current="page" class="ld-crumb__link ld-crumb__link--current" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    baz
  </ld-crumb>
</ld-breadcrumbs>
`;

exports[`ld-breadcrumbs updates current after removal of current crumb 1`] = `
<ld-breadcrumbs>
  <mock:shadow-root>
    <nav aria-label="Breadcrumbs" class="ld-breadcrumbs">
      <ol class="ld-breadcrumbs__list" part="list">
        <slot></slot>
      </ol>
    </nav>
  </mock:shadow-root>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link chevron="end" class="ld-crumb__link" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    foo
  </ld-crumb>
  <ld-crumb>
    <mock:shadow-root>
      <li class="ld-crumb">
        <ld-link aria-current="page" class="ld-crumb__link ld-crumb__link--current" part="link">
          <slot></slot>
        </ld-link>
      </li>
    </mock:shadow-root>
    bar
  </ld-crumb>
</ld-breadcrumbs>
`;
